### PR TITLE
Text Generate REST API schema

### PR DESCRIPTION
**Pull request by gavrishp**
**Propose generate rest api endpoints**

```
/v2/models/{model_name}/versions/${MODEL_VERSION}/generate
/v2/models/{model_name}/versions/${MODEL_VERSION}/generate_stream
```
[
Screenshot 2024-01-16 at 6 43 56 PM
Screenshot 2024-01-16 at 6 44 46 PM
Screenshot 2024-01-16 at 6 46 46 PM
](url)

Reference - https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/protocol/extension_generate.html#generate-extension

### CODE REVIEW

File: specification/protocol/generate_rest.yaml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          in: path
+          schema:
+            type: string
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/GenerateRequest'
+      responses:
+        '200':
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:64]: Can we also define the error responses ?

File: specification/protocol/generate_rest.yaml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    GenerateRequest:
+      type: object
+      required: 
+        - text_input
+      properties:
+        text_input:
+          type: string
+        parameters:
+          $ref: '#/components/schemas/GenerateParameters'
+    GenerateParameters:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:17]: I think we might still want to define a list of common properties and then in addition user can still add custom parameters.

File: specification/protocol/generate_rest.yaml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+          type: string
+        model_version:
+          type: string
+    GenerateStreamResponse:
+      type: object
+      required:
+        - text_output
+        - model_name
+        - model_version
+        - done
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:41]: Missing the done field on the property list

File: specification/protocol/generate_rest.yaml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      properties:
+        text_output:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:43]: This is concatenated text output, we might still want to see the token generated for each iteration.

Comment (gavrishp) [reply specification/protocol/generate_rest.yaml:43]: In the Nvidia implementation, each response in returning cumulative set of tokens.
```
1st json
{
text_output: "Here is"
}
.
.
.
..
subsequent json response
{
text_output: "Here is the output for the prompt"
}
```

Should we add additional property to display token generated in current response set?

**Comment by gavrishp** [comment]
@yuzisun Wanted to follow up, if the current state of changes are alright?

File: specification/protocol/generate_rest.yaml (hunk 1 above)
Comment (yuzisun) [reply specification/protocol/generate_rest.yaml:108]: I think we would want to define a list of http error response with the `GenerateErrorResponse` like here
https://github.com/huggingface/text-generation-inference/blob/main/docs/openapi.json#L50

File: specification/protocol/generate_rest.yaml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        - text_input
+      properties:
+        text_input:
+          type: string
+        parameters:
+          $ref: '#/components/schemas/GenerateParameters'
+    GenerateParameters:
+      type: object
+      properties:
+        parameter:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:20]: This only allows one single parameter, we can probably use allOf
```yaml
parameters:
  allOf:
    - $ref: '#/components/schemas/GenerateParameters'
    - type: object
       additionalProperties:
         type: string
```

Comment (gavrishp) [reply specification/protocol/generate_rest.yaml:20]: Screenshot 2023-12-20 at 5 15 58 PM

Just want to make sure I got it correct. Does this match the expectation?

File: specification/protocol/generate_rest.yaml (hunk 1 above)
Comment (gavrishp) [reply specification/protocol/generate_rest.yaml:108]: Thanks! will get this updated

File: specification/protocol/generate_rest.yaml (hunk 5 above)
Comment (yuzisun) [reply specification/protocol/generate_rest.yaml:20]: actually we can make the additional parameters free formed with `additionalProperties: {}`, not limited to string

File: specification/protocol/generate_rest.yaml [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+          type: number
+          format: float
+          minimum: 0
+          description: Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.
+    GenerateResponse:
+      type: object
+      required:
+        - text_output
+        - model_name
+        - model_version
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:62]: this should be optional

File: specification/protocol/generate_rest.yaml [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    GenerateParameters:
+        allOf:
+          $ref: '#/components/schemas/GenerateParameters'
+        type: object
+        additionalProperties: {}
+        properties:
+          temperature:
+            type: number
+            format: float
+            default: null
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:26]: vllm defaults to 1
https://github.com/vllm-project/vllm/blob/main/vllm/sampling_params.py#L100

File: specification/protocol/generate_rest.yaml [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            type: number
+            format: float
+            maximum: 1
+            minimum: 0
+            description: An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.
+          max_tokens:
+            type: integer
+            format: int32
+            default: 20
+            minimum: 0
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:39]: min is at least 1
https://github.com/vllm-project/vllm/blob/main/vllm/sampling_params.py#L186

File: specification/protocol/generate_rest.yaml [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            maximum: 512
+            description: The maximum number of tokens to generate in the completion.
+          stop:
+            type: array
+            items:
+              type: string
+            maxItems: 5
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:46]: does not seem like there is max defined 
https://github.com/vllm-project/vllm/blob/main/vllm/sampling_params.py#L69

File: specification/protocol/generate_rest.yaml (hunk 8 above)
Comment (gavrishp) [reply specification/protocol/generate_rest.yaml]: 512 as upper limit is good?

File: specification/protocol/generate_rest.yaml (hunk 7 above)
Comment (gavrishp) [reply specification/protocol/generate_rest.yaml]: Updated

File: specification/protocol/generate_rest.yaml (hunk 9 above)
Comment (gavrishp) [reply specification/protocol/generate_rest.yaml]: Updated

File: specification/protocol/generate_rest.yaml (hunk 8 above)
Comment (gavrishp) [reply specification/protocol/generate_rest.yaml]: resolved

**Comment by cmaddalozzo** [comment]
We should probable add the option to return log probabilities in the result. This seems to be fairly common among other APIs. This would comprise a boolean `logprobs` parameter in the request and a corresponding `logprobs` property in the response containing an array of objects with keys `token` and `logprob`.

File: specification/protocol/generate_rest.yaml [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+          type: string
+        model_name:
+          type: string
+        model_version:
+          type: string
+        finish_reason:
+          type: string
+        logprobs:
+          $ref: '#/components/schemas/Logprobs'
+    Logprobs:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:84]: suggest change the naming to `Token`, see https://github.com/huggingface/text-generation-inference/blob/main/docs/openapi.json#L844. For streaming case it is a single token.

File: specification/protocol/generate_rest.yaml [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        - text_output
+        - model_name
+      properties:
+        text_output:
+          type: string
+        model_name:
+          type: string
+        model_version:
+          type: string
+        logprobs:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:66]: suggest change the name here, in TGI it is called `details` which includes the `tokens`, not sure if we should follow the same.
https://github.com/huggingface/text-generation-inference/blob/main/docs/openapi.json#L844

File: specification/protocol/generate_rest.yaml [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+      properties:
+        text_output:
+          type: string
+        model_name:
+          type: string
+        model_version:
+          type: string
+        finish_reason:
+          type: string
+        logprobs:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:82]: For streaming case it is a single token.

File: specification/protocol/generate_rest.yaml (hunk 11 above)
Comment (yuzisun) [reply specification/protocol/generate_rest.yaml:66]: in OpenAI it is `choices`
https://platform.openai.com/docs/api-reference/chat/object

File: specification/protocol/generate_rest.yaml [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+      type: object
+      required: 
+        - text_input
+      properties:
+        text_input:
+          type: string
+        parameters:
+          allOf: 
+            - $ref: '#/components/schemas/GenerateParameters'
+        logprob:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:52]: should this be part of the `GenerateParameters` ?

File: specification/protocol/generate_rest.yaml [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+          default: 20
+          minimum: 1
+          description: The maximum number of tokens to generate in the completion.
+        stop:
+          type: array
+          items:
+            type: string
+          description: Sequences where the API will stop generating further tokens.
+        logprob:
+          type: boolean
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:51]: Can you add a description for this flag?

File: specification/protocol/generate_rest.yaml [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+      properties:
+        text_output:
+          type: string
+        model_name:
+          type: string
+        model_version:
+          type: string
+        details:
+          $ref: '#/components/schemas/StreamDetails'
+    Logprobs:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:90]: Add a description for this

File: specification/protocol/generate_rest.yaml [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+          type: integer
+          format: int32
+          minimum: 0
+        logprob:
+          type: number
+          format: float
+        special:
+          type: boolean
+        text:
+          type: string
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:120]: Let's make sure we have descriptions for these fields

File: specification/protocol/generate_rest.yaml [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+  title: Open Inference API for text generation
+  description: Open Inference API for text generation
+  version: 1.0.0
+components:
+  schemas:
+    Details:
+      type: object
+      additionalProperties: {}
+      properties:
+        finish_reason:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:12]: finish_reason should be an enum

File: specification/protocol/generate_rest.yaml [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+          type: string
+        logprobs:
```
Comment (yuzisun) [inline specification/protocol/generate_rest.yaml:14]: both `finish_reason` and `logprobs` should be required if details is requested.

**Comment by yuzisun** [comment]
Thanks @gavrishp !! Great job on getting this going with the initial version.

/lgtm
/approve

**Pull request merged**